ddl/db_test.go

```diff
@@ -4514,6 +4514,136 @@ func (s *testSerialDBSuite) TestModifyColumnCharset(c *C) {
+
+	// to date/datetime/timestamp
+	tk.MustExec("create table t_mctd2(a time, b time, c time)")
+	tk.MustExec(`insert into t_mctd2 (a,b,c) values
+("2:8:12", "2:8:12", "2:8:12"),
+("20:00:12", "20:00:12", "20:00:12"),
+("30 20:00:12", "30 20:00:12", "30 20:00:12"),
+("000012", "000012", "000012"),
+("002012.4358", "002012.4358", "002012.4358"),
+("28 000012.4358", "28 000012.4358", "28 000012.4358"),
+(000000, 000000, 000000),
+(000000.549, 000000.549, 000000.549)
+`)
+	tk.MustExec(`alter table t_mctd2 modify column a date`)
+	tk.MustExec(`alter table t_mctd2 modify column b datetime`)
+	tk.MustExec(`alter table t_mctd2 modify column c timestamp`)
+	result = tk.MustQuery("select * from t_mctd2")
+
+	buildDate := func(t time.Time) string {
+		return fmt.Sprintf("%d-%02d-%02d", t.Year(), t.Month(), t.Day())
+	}
+	buildTime := func(t time.Time) string {
+		return fmt.Sprintf("%02d:%02d:%02d", t.Hour(), t.Minute(), t.Second())
+	}
+	buildRow := func(t time.Time, hour, minute, second int) string {
+		t = time.Date(t.Year(), t.Month(), t.Day(), hour, minute, second, 0, time.UTC)
+		return fmt.Sprintf("%s %s %s %s %s", buildDate(t), buildDate(t), buildTime(t), buildDate(t.Local()), buildTime(t.Local()))
+	}
+
+	result.Check(testkit.Rows(
+		buildRow(time.Now().UTC(), 2, 8, 12),
+		buildRow(time.Now().UTC(), 20, 0, 12),
+		buildRow(time.Now().UTC().AddDate(0, 0, 30), 20, 0, 12),
+		buildRow(time.Now().UTC(), 0, 0, 12),
+		buildRow(time.Now().UTC(), 0, 20, 12),
+		buildRow(time.Now().UTC().AddDate(0, 0, 28), 12, 0, 0),
+		buildRow(time.Now().UTC(), 0, 0, 0),
+		buildRow(time.Now().UTC(), 0, 0, 1),
+	))
+
+	// to time
+	tk.MustExec("create table t_mctt1(a year)")
+	defer tk.MustExec("drop table t_mctt1")
+	tk.MustExec(`insert into t_mctt1 (a) values ("07")`)
+	tk.MustExec(`alter table t_mctt1 modify column a time`)
+	result = tk.MustQuery("select * from t_mctt1")
+	result.Check(testkit.Rows("00:20:07"))
+
+	tk.MustExec("create table t_mctt2(a date, b datetime, c timestamp)")
+	defer tk.MustExec("drop table t_mctt2")
+	tk.MustExec(`insert into t_mctt2 (a, b, c) values
+("2019-09-12", "2019-09-12 11:40:30", "2019-09-12 11:40:30"),
+("18-09-12", "18-09-12 10:30:55", "18-09-12 10:30:55"),
+("20170912", "20170912234012", "20170912234012"),
+("160912", "160912184752", "160912184752"),
+("000912", "000912150333", "000912150333"),
+("700912", "700912174928", "700912174928"),
+(20190912, 20190912114030, 20190912114030),
+(180912, 180912103055, 180912103055),
+(700912, 700912234012, 700912234012)
+`)
+	tk.MustExec(`alter table t_mctt2 modify column a time`)
+	tk.MustExec(`alter table t_mctt2 modify column b time`)
+	tk.MustExec(`alter table t_mctt2 modify column c time`)
+	result = tk.MustQuery("select * from t_mctt2")
+
+	buildRow2 := func(hour, minute, second int) string {
+		t := time.Date(2020, 0, 0, hour, minute, second, 0, time.Local)
+		return fmt.Sprintf("00:00:00 %s %s", buildTime(t), buildTime(t.UTC()))
+	}
+	result.Check(testkit.Rows(
```

Comment:
add some test for datetime(fsp) / timestamp(fsp)

Reply:
There is a full cover test on this PR, which I have not merged into this PR.
The thing is lots of behaviors are not aligned to mysql. I am carrying up the process today.

Reply:
great! can you make a list recording these incompatible behaviors, which should be occurred in our docs as well

Reply:
Sorry for late resp. I've fixed the test, and integrated it into the newest commit.

You can automatically test the listed tests with mysql/mariadb, just like I did by the code in comment https://github.com/pingcap/tidb/pull/19988#issuecomment-696864223.

And all tests(in db_test.go) passed on mariadb. Those failed on TiDB db_test.go are marked with `//`, or commented. There are some useful comments you may need.

I forbid all cases that may fail. Since I think it is a long time since the open of this PR. We better merge it now...